File: cinder/Entrainment/src/ViewTerrain.cpp

//
//  ViewTerrain.cpp
//  Entrainment03
//
//  Created by Yi-Wen Lin on 2015/7/26.
//
//

#include "ViewTerrain.h"
#include "GlobalSettings.h"
#include "MeshUtils.h"

using namespace ci;
using namespace ci::app;
using namespace std;

ViewTerrain::ViewTerrain() : View("shaders/terrain.vert", "shaders/terrain.frag") {
}

ViewTerrain::ViewTerrain(float mNumSeg, float mSize, float mHeight) : View("shaders/terrain.vert", "shaders/terrain.frag"), _numSeg(mNumSeg), _size(mSize), _height(mHeight) {
    _init();
}


Vec3f ViewTerrain::_getPosition(int i, int j, float numSeg) {
    float sx = -_size * .5;
    Vec3f pos;
    pos.x = sx + i/_numSeg * _size;
    pos.z = sx + j/_numSeg * _size;
    
    return pos;
}


void ViewTerrain::_init() {
    gl::VboMesh::Layout     _layout;
    _layout.setStaticIndices();
    _layout.setStaticPositions();
    _layout.setStaticTexCoords2d();
    
    vector<Vec3f> positions;
    vector<Vec2f> coords;
    vector<uint> indices;
    int count = 0;
    float uvGap = 1.0 / _numSeg;
    
    for(int j=0; j<_numSeg; j++) {
        for(int i=0; i<_numSeg; i++) {
            positions.push_back(_getPosition(i, j+1, _numSeg));
            positions.push_back(_getPosition(i+1, j+1, _numSeg));
            positions.push_back(_getPosition(i+1, j, _numSeg));
            positions.push_back(_getPosition(i, j, _numSeg));
            
            coords.push_back(Vec2f(i/_numSeg, j/_numSeg+uvGap));
            coords.push_back(Vec2f(i/_numSeg+uvGap, j/_numSeg+uvGap));
            coords.push_back(Vec2f(i/_numSeg+uvGap, j/_numSeg));
            coords.push_back(Vec2f(i/_numSeg, j/_numSeg));
            
            indices.push_back(count * 4 + 0);
            indices.push_back(count * 4 + 1);
            indices.push_back(count * 4 + 2);
            indices.push_back(count * 4 + 0);
            indices.push_back(count * 4 + 2);
            indices.push_back(count * 4 + 3);
            
            count ++;
        }
    }
    
    
    mesh = gl::VboMesh(positions.size(), indices.size(), _layout, GL_TRIANGLES);
    mesh.bufferPositions(positions);
    mesh.bufferTexCoords2d(0, coords);
    mesh.bufferIndices(indices);
}

void ViewTerrain::render(gl::Texture texture, gl::Texture textureNormal, gl::Texture textureBump, Vec3f pos, Vec2f uvOffset, float numTiles, Vec3f lightPos, Vec3f eyePos) {
    shader->bind();
    shader->uniform("texture", 0);
    shader->uniform("textureNormal", 1);
    shader->uniform("textureBump", 2);
    shader->uniform("position", pos);
    shader->uniform("uvOffset", uvOffset);
    shader->uniform("numTiles", numTiles);
    shader->uniform("height", _height);
    shader->uniform("lightPos", lightPos);
    shader->uniform("eyePos", eyePos);
    shader->uniform("bumpOffset", GlobalSettings::getInstance().bump);
    shader->uniform("noiseSandUVScale", GlobalSettings::getInstance().noiseSandUVScale);
    shader->uniform("specularLightStrength", GlobalSettings::getInstance().specularLightStrength);
    shader->uniform("diffuseLightStrength", GlobalSettings::getInstance().diffuseLightStrength);
    shader->uniform("diffuseColor", GlobalSettings::getInstance().colorDiffuse);
    shader->uniform("spotLightShiness", GlobalSettings::getInstance().spotLightShiness);
    shader->uniform("spotlightRadius", GlobalSettings::getInstance().spotLightRadius);
    shader->uniform("spotLightStrength", GlobalSettings::getInstance().spotLightStrength);
    shader->uniform("spakleStrength", GlobalSettings::getInstance().sparkleStrength);
    texture.bind(0);
    textureNormal.bind(1);
    textureBump.bind(2);
    gl::draw(mesh);
    texture.unbind();
    textureNormal.unbind();
    textureBump.unbind();
    shader->unbind();
}

File: webgl/src/js/terrain/ViewTerrain.js
// ViewTerrain.js

var GL = bongiovi.GL;
var gl;
var glslify = require("glslify");

function ViewTerrain(mSegs, mSize, mHeight) {
	this._height = mHeight || 0.0;
	console.log(this._height);
	this._size   = mSize;
	this._numSeg = mSegs;
	bongiovi.View.call(this, glslify("../../shaders/terrain.vert"), glslify("../../shaders/terrain.frag"));
	gl = GL.gl;
	// new TangledShader(gl, this.shader.fragmentShader, this._onShaderUpdate.bind(this));
}

var p = ViewTerrain.prototype = new bongiovi.View();
p.constructor = ViewTerrain;


p._init = function() {
	gl            = GL.gl;
	var positions = [];
	var coords    = [];
	var indices   = []; 
	var index     = 0;
	var size      = this._size;
	var uvgap     = 1/this._numSeg;

	function getPos(i, j, num) {
		var sx = -size * .5;
		var x = sx + i/num * size;
		var z = sx + j/num * size;
		return [x, 0, z];
	}


	for(var j=0; j<this._numSeg; j++) {
		for(var i=0; i<this._numSeg; i++) {
			positions.push(getPos(i, j+1, this._numSeg));
			positions.push(getPos(i+1, j+1, this._numSeg));
			positions.push(getPos(i+1, j, this._numSeg));
			positions.push(getPos(i, j, this._numSeg));

			coords.push([i/this._numSeg, j/this._numSeg+uvgap]);
			coords.push([i/this._numSeg+uvgap, j/this._numSeg+uvgap]);
			coords.push([i/this._numSeg+uvgap, j/this._numSeg]);
			coords.push([i/this._numSeg, j/this._numSeg]);

			indices.push(index * 4 + 0);
			indices.push(index * 4 + 1);
			indices.push(index * 4 + 2);
			indices.push(index * 4 + 0);
			indices.push(index * 4 + 2);
			indices.push(index * 4 + 3)

			index++;
		}
	}

	this.mesh = new bongiovi.Mesh(positions.length, indices.length, GL.gl.TRIANGLES);
	this.mesh.bufferVertex(positions);
	this.mesh.bufferTexCoords(coords);
	this.mesh.bufferIndices(indices);
};


p._onShaderUpdate = function(shader) {
	this.shader.attachShaderProgram();
};


p.render = function(texture, textureNormal, textureBump, pos, uvOffset, numTiles, lightPos, eyePos, spotLightPos) {
	this.shader.bind();
	this.shader.uniform("texture", "uniform1i", 0);
	this.shader.uniform("textureNormal", "uniform1i", 1);
	this.shader.uniform("textureBump", "uniform1i", 2);
	this.shader.uniform("position", "uniform3fv", pos);
	this.shader.uniform("uvOffset", "uniform2fv", uvOffset);
	this.shader.uniform("numTiles", "uniform1f", numTiles);
	this.shader.uniform("height", "uniform1f", this._height);
	this.shader.uniform("lightPos", "uniform3fv", lightPos);
	this.shader.uniform("spotLightPos", "uniform3fv", spotLightPos);
	this.shader.uniform("eyePos", "uniform3fv", [ -eyePos[0], eyePos[1], eyePos[2]]);
	this.shader.uniform("bumpOffset", "uniform1f", params.bump);
	this.shader.uniform("spotlightRadius", "uniform1f", params.spotlightRadius);
	this.shader.uniform("spotLightShiness", "uniform1f", params.spotLightShiness);
	this.shader.uniform("noiseSandUVScale", "uniform1f", params.noiseSandUVScale);
	this.shader.uniform("showSpecularLight", "uniform1f", params.showSpecularLight ? 1.0 : 0.0);
	this.shader.uniform("showGlitterLight", "uniform1f", params.showGlitterLight ? 1.0 : 0.0);
	this.shader.uniform("specularLightStrendght", "uniform1f", params.specularLightStrendght);
	this.shader.uniform("spotLightStrength", "uniform1f", params.spotLightStrength);
	this.shader.uniform("diffuseLightStrendght", "uniform1f", params.diffuseLightStrendght);
	texture.bind(0);
	textureNormal.bind(1);
	textureBump.bind(2);
	GL.draw(this.mesh);
};

module.exports = ViewTerrain;